dlx2.py

```diff
@@ -0,0 +1,178 @@
+
+class Column:
+    def __init__(self, header):
+        self.head = header
+        self.tail = header
+        self.right = self
+        self.left = self
+        self.size = 1
+
+    def add(self, node):
+        node.up = self.tail
+        node.down = self.tail.down
+        self.tail.down = node
+        self.tail = node
+        self.head.up = node
+        self.size += 1
+
+    def __repr__(self):
+        string = ''
+        current_node = self.head
+        for i in range(self.size):
+            string += str(current_node)
+            current_node = current_node.down
+        return string
+
+class ColumnList:
+    def __init__(self, head_col):
+        self.head_col = head_col
+        self.tail_col = head_col
+        self.size = 1
+
+    def add(self, col):
+        col.right = self.tail_col.right
+        col.left = self.tail_col
+        self.tail_col.right = col
+        self.tail_col = col
+        self.head_col.left = col
+        self.size += 1
+
+    def __repr__(self):
+        string = ''
+        current_col = self.tail_col
+    
+        for i in range(self.size):
+            string += str(current_col)
+            string += '\n'
+            current_col = current_col.left
+        
+        return string
+
+# Class for a sudoku cover matrix 729 by 324
+# Each row represents a choice (e.g. 4 in cell 9, 5 in cell 80, 9 in cell 32, etc.)
+# Rows follow this order: cell 1 (guesses 1-9), cell 2 (guesses 1-9), etc.
+# Columns are used to eliminate conflicting choices
+# Columns 1-81 ensure that all other choices for a cell are eliminated after the first choice has been made
+# Columns 82-162 ensure that a digit may only be guessed once per sudoku row
+# Columns 163-243 ensure that a digit may only be guessed once per sudoku column
+# Columns 244-324 ensure that a digit may only be guessed once per box
+class Cover:
+    def __init__(self):
+        # initializes empty matrix
+        self.matrix = [[0 for j in range(81 * 4)] for i in range(9 * 81)]
+        # loops through rows to assign constraints
+        for i, row in enumerate(self.matrix):
+            # cells change every 9 rows
+            cell_number = int((i - (i % 9))/9)
+            # sudoku rows change every 81 rows
+            row_number = int((i - (i % 81))/81)
+            # sudoku columns change for every cell and reset every 9 cells
+            col_number = int(cell_number % 9)
+            # sudoku boxes are calculated by taking the lower row of the box (0,3, or 6) plus the column grouping (1,2, or 3)
+            # result is one of nine boxes (0 through 8)
+            box_number = int((row_number - (row_number % 3)) + ((col_number - (col_number % 3))/3))
+
+            # cell constraint gets placed in the corresponding column for the cell number 1-81
+            cell_i = cell_number
+            row[cell_i] = Node(1, i, cell_i)
+
+            # row constraint gets placed in 9 by 9 diagonals that are offset by 9 times the row number (0 through 8)
```

Comment:
I was curious to see what this looks like in practice so I threw together a quick python script to visualize it. Visually matches how you describe it:
![sudoku-cover-matrix](https://user-images.githubusercontent.com/8422414/128640494-0d3844b9-bfa2-4340-b959-99a9d24bdbfb.png).

Here's the script (zipped because GH doesn't support uploading raw code files). Requires `numpy` (very useful matrix math library) and `matplotlib` (plotting library)
[sudoku_visualization.zip](https://github.com/carlschr/python-command-line-sudoku/files/6951215/sudoku_visualization.zip)